expan/core/experiment.py

```
@@ -1,280 +1,232 @@
-        self.reference_kpis       = {}
-
-        # add derived KPIs to the data frame
-        for name, formula in zip(derived_kpi_names, derived_kpi_formulas):
-            self.data.loc[:, name] = eval(re.sub(kpi_name_pattern, r'self.data.\1.astype(float)', formula))
-            self.reference_kpis[name] = re.sub(kpi_name_pattern + '/', '', formula)
-
-    def get_kpi_by_name_and_variant(self, data, name, variant):
-        return data.reset_index().set_index('variant').loc[variant, name]
+    """ Class which adds the analysis functions to experimental data. """
+    def __init__(self, data, metadata):
+        self.data         = data.copy()
+        self.metadata     = metadata.copy()
+        self.worker_table = {
+            'fixed_horizon': statx.make_delta,
+            'group_sequential': es.make_group_sequential,
+            'bayes_factor': es.make_bayes_factor,
+            'bayes_precision': es.make_bayes_precision
+        }
+
 
     def __str__(self):
-        variants = self.variant_names
+        return 'Experiment "{:s}" with {:d} entities.'.format(self.metadata['experiment'], len(self.data))
+
+
+    def analyze_statistical_test(self, test, testmethod, **worker_args):
+        """ Runs delta analysis on one statistical test and returns statistical results.
+        
+        :param test: a statistical test to run
+        :type  test: StatisticalTest
+        :param testmethod: analysis method
+        :type  testmethod: str
+        :param **worker_args: additional arguments for the analysis method
+        :return: statistical result of the test
+        :rtype: StatisticalTestResult
+        """
+        if not isinstance(test, StatisticalTest):
+            raise TypeError("Statistical test should be of type StatisticalTest.")
+
+        if 'entity' not in self.data.columns():
+            raise RuntimeError("There is no 'entity' column in the data.")
+        if self.data.entity.duplicated().any():
```

I think that it used to be a concern of Expan to filter out those duplicates. Am I wrong?

---

yes it is used to filter out duplicates. Should we still do the same here?

---

I think that it's quite often the case that data has duplicates — I think we shouldn't break things that we don't have to break and keep at least a resemblance of backward compatibility in terms of functionality.

---

oh I think I misunderstood.

Here ```entity``` should not have duplicates because we do our analysis based on this entity level. If there is duplicates, something is wrong when we group the data. And I think we should raise an error if this happens.